### ISSUE TITLE
using `sum()` causes dtype mismatch

### ISSUE THREAD

**Issue by JackTemaki** [rwth-i6/returnn_common#135]
Writing e.g. `l2loss = sum(nn.reduce(param ** 2.0, axis=list(param.shape), mode="sum") for param in net.parameters())`
Causes a dtype mismatch when running the network (not during construction with `returnn_common` surprisingly):
```python
  File "/work/asr4/rossenbach/sisyphus_work_folders/tts_asr_2021_work/i6_core/tools/git/CloneGitRepositoryJob.UbdRLURNoRpi/output/repository/returnn/tf/util/data.py", line 3614, in Data.copy_compatible_to
    line: assert not check_dtype or self.dtype == data.dtype
    locals:
      check_dtype = <local> True
      self = <local> Data{'reduce_output', []}
      self.dtype = <local> 'float32', len = 7
      data = <local> Data{'add_output', [], dtype='int32'}
      data.dtype = <local> 'int32
```
The error on the Returnn side is justified, because the generated layers are:
```python
    "reduce": {
        "class": "reduce",
        "from": "pow",
        "mode": "sum",
        "axis": [transformer_default_model_dim, _3_transformer_default_model_dim],
        "out_shape": {},
    },
    "constant_0": {"class": "constant", "value": 0},
    "add": {
        "class": "combine",
        "from": ["constant_0", "reduce"],
        "kind": "add",
        "out_shape": {},
    },
```
As the value is `0` and not `0.0` Returnn assumes int32 instead of float32 (float32 is the correct type of the `reduce` output). The `0` is added probably on the first `sum()` internal call of:
```python
  def __add__(self, other: Union[RawTensorTypes, Tensor]) -> Tensor:
    from ._generated_layers import _combine
    return _combine([self, nn.convert_to_tensor(other)], kind="add", name="add")
```

I am not sure how to deal with the correct dtypes here, as when calling `sum()` we do may not exactly know which dtype we want to have.

One thing I could image is that we infer the type from `self` if `other` is a zero, and extend the `nn.convert_to_tensor` to have a dtype parameter.

**Comment by albertz** [comment]
Yea, `sum([p1, p2])` is equivalent to `0 + p1 + p2`.
In TF and PyTorch int + float is allowed and automatically upcasted.
In RETURNN, in the `CombineLayer`, I think we don't allow this (I assume this is your error; you have not postet the full traceback).

So, obviously we cannot change how `sum` is defined, as this is defined by Python itself.

We could add an optimization on returnn-common side which automatically reduces `0 + x` or `x + 0` to `x`, no matter the type of `x`.

We could extend RETURNN `CombineLayer` to allow int + float.

Not sure which way is the best or you would prefer. I think the returnn-common optimization would be nice anyway and would already solve this.

**Comment by albertz** [comment]
> we do may not exactly know which dtype we want to have.

Note that such upcasting (promotion) rules are well defined in PyTorch/TF.

https://pytorch.org/docs/stable/tensor_attributes.html#torch.torch.dtype

**Comment by JackTemaki** [comment]
Okay then lets start with the optimizations like x+0 -> x; x-0 -> x; x*1 -> x and so on... I think those are a good idea anyway.

**Comment by albertz** [comment]
Ok done.

**Comment by albertz** [comment]
Btw, if you want to have an L2 loss, I would not recommend to do it this way. Rather, see #59. And you should also enable `decouple_constraints` then.
